src/relay/transforms/defunctionalization.cc

```diff
@@ -0,0 +1,432 @@
+ * In addition, our function argument `(\x -> \x + 1)` will be encoded as a datatype constructor,
+ * which we will call `incr`, and all calls to `F` in our specialized map function will use the
+ * helper `apply` function.
+ *
+ * After defunctionalization, we get:
+ * fun apply encoding arg =  case encoding of
+ *     “incr” => incr arg
+ * fun map’ F y = case y of
+ *           Nil => Nil
+ *           | Cons(x, xs) => Cons(apply F x, map’ F xs)
+ * fun addone 1 = map’ “incr” 1
+ *
+ * Currently, defunctionalization makes the following assumptions:
+ * - functions cannot return function values
+ * - function arguments are in two forms: identifier or a lambda abstraction
+ * - no functions stored in datatype
+ * - functions are not let binded
+ */
+
+#include <tvm/ir/type_functor.h>
+#include <tvm/relay/analysis.h>
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/feature.h>
+#include <tvm/relay/transform.h>
+#include <tvm/te/operation.h>
+
+#include "../analysis/type_solver.h"
+#include "../transforms/pass_util.h"
+namespace tvm {
+namespace relay {
+
+// determine if type contains a FuncType
+bool HasFuncType(const Type& t) {
+  struct FuncTypeVisitor : TypeVisitor {
+    bool has_func_type;
+    FuncTypeVisitor() : has_func_type(false) {}
+
+    void VisitType_(const FuncTypeNode* op) { this->has_func_type = true; }
+  };
+
+  auto visitor = FuncTypeVisitor();
+  visitor.VisitType(t);
+  return visitor.has_func_type;
+}
+// determine if FuncType is a higher order type
+bool IsHigherOrderFunc(const FuncType& t) {
+  bool higher_order = false;
+  for (auto arg : t->arg_types) {
+    higher_order |= HasFuncType(arg);
+  }
+  return higher_order |= HasFuncType(t->ret_type);
+}
+
+/*!
+ * \brief mutator for driving the Defunctionalization transformation
+ */
+class DefuncMutator : public ExprMutator {
+ public:
+  explicit DefuncMutator(const IRModule& mod) : mod(mod), constructor_counter(0) {}
+
+  Expr VisitExpr_(const CallNode* call) {
+    if (auto op = call->op.as<GlobalVarNode>()) {
+      CHECK_EQ(call->type_args.size(), op->checked_type().as<FuncTypeNode>()->type_params.size())
+          << "all type args must be explicit";
+
+      auto op_type = InstFuncType(op->checked_type().as<FuncTypeNode>(), call->type_args);
+      CHECK_EQ(FreeTypeVars(op_type, mod).size(), 0) << "free type vars in instantiated";
+      CHECK(!HasFuncType(op_type->ret_type)) << "returning functions not supported";
+
+      if (!IsHigherOrderFunc(op_type)) {
+        // not higher order function
+        return ExprMutator::VisitExpr_(call);
+      }
+
+      // first we encode function arguments
+      Array<Expr> args;
+      for (size_t i = 0; i < call->args.size(); i++) {
+        auto arg = call->args[i];
+        auto type = op_type->arg_types[i];
+        if (!HasFuncType(type)) {
+          args.push_back(arg);
+          continue;
+        }
+
+        args.push_back(EncodeArg(arg, type));
```

Comment:
if else is a better style then continue.

Reply:
yep, looks like CI issue was caused by this: https://github.com/apache/incubator-tvm/pull/6434